presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java

```diff
@@ -0,0 +1,471 @@
+import com.facebook.presto.execution.ScheduledSplit;
+import com.facebook.presto.execution.SplitRunner;
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.execution.TaskSource;
+import com.facebook.presto.execution.TaskStateMachine;
+import com.facebook.presto.execution.executor.TaskExecutor;
+import com.facebook.presto.execution.executor.TaskHandle;
+import com.facebook.presto.operator.Driver;
+import com.facebook.presto.operator.DriverContext;
+import com.facebook.presto.operator.DriverFactory;
+import com.facebook.presto.operator.DriverStats;
+import com.facebook.presto.operator.PipelineContext;
+import com.facebook.presto.operator.TaskContext;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.LocalExecutionPlanner.LocalExecutionPlan;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.Duration;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.facebook.presto.SystemSessionProperties.getInitialSplitsPerNode;
+import static com.facebook.presto.SystemSessionProperties.getMaxDriversPerTask;
+import static com.facebook.presto.SystemSessionProperties.getSplitConcurrencyAdjustmentInterval;
+import static com.facebook.presto.operator.PipelineExecutionStrategy.UNGROUPED_EXECUTION;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+/**
+ * The PrestoSparkTaskExecution is a simplified version of SqlTaskExecution.
+ * It doesn't support grouped execution that is not needed on Presto on Spark.
+ * Unlike the SqlTaskExecution the PrestoSparkTaskExecution does not require
+ * the output buffer to be drained to mark the task as finished. As long as
+ * all driver as finished the task execution is marked as finished. That allows to
+ * have more control over the output Iterator lifecycle in the PrestoSparkTaskExecutor
+ */
+public class PrestoSparkTaskExecution
```

Comment:
Are you forking the current `SqlTaskExecution` (and remove grouped execution related stuff), or you are forking the old `SqlTaskExecution` (i.e. before commit 4e55aadca50e2b6f904fcfa444540e90c02e3383) . In my opinion the code is much easier to follow with the old `SqlTaskExecution`

Reply:
I took the current version of `SqlTaskExecution` as a base, but I was looking at the old version while removing grouped execution code.